### PR TITLE
TPM registration over AMQP

### PR DESCRIPTION
**Pull request by BertKleewein**
I'm getting this code into the review pipeline to avoid the rush.  

This code has several prerequisites:
1. Pierre's AMQP changes.  This code will need to be edited for the new API.
2. Tony's TPM commit.  I will need to re-merge this code on top of Tony's
3. My previous review -- PR 208 (prep work for TPM over AMQP).  I'm waiting for Tony's TPM changes before I commit this.

After this gets committed, I have 3 things I want to do before I consider my DPS work complete:
1. Increase code coverage numbers, particularly for the AMQP transport, and maybe the provisioning device client.
2. Add error event handler to the AMQP provisioning transport.
3. Update Jenkins to use the TPM simulator so we can run TPM E2E in the gate.  (unless someone else wants to volunteer for this. :)

---
This change is [Reviewable](https://reviewable.io/reviews/azure/azure-iot-sdk-node/212)

### CODE REVIEW

**Comment by pierreca** [comment]
The _provisioning_e2e.js file is getting really long I wonder if it might be a good thing to break it down and "extract" the helper functions in separate files 

---

Reviewed 13 of 13 files at r1.
Review status: all files reviewed at latest revision, 19 unresolved discussions.

---

*[common/transport/amqp/src/amqp.ts, line 402 at r1](https://reviewable.io:443/reviews/azure/azure-iot-sdk-node/212#-L3pLv7BR7yOBeOXfToH:-L3pLv7BR7yOBeOXfToI:b-n8k0ur) ([raw file](https://github.com/azure/azure-iot-sdk-node/blob/1104e870439d2345151b143ac00efad5122411e5/common/transport/amqp/src/amqp.ts#L402)):*
> ```XML
>    * @param {Function}   done          Callback called when the connection is established or if an error happened.
>    */
>   connect(uri: string, sslOptions: any, done: GenericAmqpBaseCallback<any>): void {
> ```

just wondering, have you considered changing the purpose of the sslOptions object instead of adding another method? that'd be a breaking change but it's not really a problem on internal packages, and it would keep the API "cleaner"? 

---

*[common/transport/amqp/src/amqp.ts, line 570 at r1](https://reviewable.io:443/reviews/azure/azure-iot-sdk-node/212#-L3pMFNDKK1Pd12-0tGf:-L3pMFNDKK1Pd12-0tGg:b-wbullh) ([raw file](https://github.com/azure/azure-iot-sdk-node/blob/1104e870439d2345151b143ac00efad5122411e5/common/transport/amqp/src/amqp.ts#L570)):*
> ```XML
>   }
> 
>   private _setUri(uri: string): void {
> ```

nit: setUriAndSocketTransport ? after all it's not only about the URI..

---

*[common/transport/mqtt/package.json, line 11 at r1](https://reviewable.io:443/reviews/azure/azure-iot-sdk-node/212#-L3pLj-MAAZWGfe-hvKC:-L3pLj-MAAZWGfe-hvKD:b-9sl6cl) ([raw file](https://github.com/azure/azure-iot-sdk-node/blob/1104e870439d2345151b143ac00efad5122411e5/common/transport/mqtt/package.json#L11)):*
> ```JSON
>   "dependencies": {
>     "azure-iot-common": "1.3.0",
>     "debug": "^3.1.0",
> ```

are we really  missing this or did this sneak up in the review inadvertantly?

---

*[provisioning/e2e/_provisioning_e2e.js, line 31 at r1](https://reviewable.io:443/reviews/azure/azure-iot-sdk-node/212#-L3pPXgJW6iRTw7pNu4C:-L3pPXgJW6iRTw7pNu4D:b-dsj0eu) ([raw file](https://github.com/azure/azure-iot-sdk-node/blob/1104e870439d2345151b143ac00efad5122411e5/provisioning/e2e/_provisioning_e2e.js#L31)):*
> ```JavaScript
> var registry = Registry.fromConnectionString(registryConnectionString);
> 
> // avoid unused object warning when commenting out tests
> ```

this is also going to avoid warning when removing tests and forgetting to remove require statements. it's not very important... why not rely on .skip() instead of commenting tests?

---

*[provisioning/e2e/_provisioning_e2e.js, line 147 at r1](https://reviewable.io:443/reviews/azure/azure-iot-sdk-node/212#-L3pSeWQ7HDlg0YBttRn:-L3pSeWQ7HDlg0YBttRo:b-hesc8r) ([raw file](https://github.com/azure/azure-iot-sdk-node/blob/1104e870439d2345151b143ac00efad5122411e5/provisioning/e2e/_provisioning_e2e.js#L147)):*
> ```JavaScript
> 
>   this.register = function (Transport, callback) {
>     var securityClient = new X509Security(this.registrationId, self._cert);
> ```

if you use this here, aren't you going to need to bind(this) later? why not use self since you seem to have it already?

---

*[provisioning/transport/amqp/devdoc/sasl_tpm_requirements.md, line 18 at r1](https://reviewable.io:443/reviews/azure/azure-iot-sdk-node/212#-L3pUku-xiA77J3iWmqT:-L3pUku-xiA77J3iWmqU:b3m2282) ([raw file](https://github.com/azure/azure-iot-sdk-node/blob/1104e870439d2345151b143ac00efad5122411e5/provisioning/transport/amqp/devdoc/sasl_tpm_requirements.md#L18)):*
> ```Markdown
> 
> 
> ## getInitFrameContent(): Promise<any>;
> ```

this has been renamed to getInitFrame()

---

*[provisioning/transport/amqp/devdoc/sasl_tpm_requirements.md, line 26 at r1](https://reviewable.io:443/reviews/azure/azure-iot-sdk-node/212#-L3pUnZ8RwR8wD75cVn3:-L3pUnZ8RwR8wD75cVn4:bbftbso) ([raw file](https://github.com/azure/azure-iot-sdk-node/blob/1104e870439d2345151b143ac00efad5122411e5/provisioning/transport/amqp/devdoc/sasl_tpm_requirements.md#L26)):*
> ```Markdown
> 
> 
> ## getChallengeResponseContent(challenge: any): Promise<Buffer>;
> ```

this has been renamed to getResponseFrame(). Also the result of the promise must be an object that contains a response field that contains the Buffer, in other words the return type is:
```ts
type SaslResponseFrame = { response: <Buffer> };
Promise<SaslResponseFrame>
```

---

*[provisioning/transport/amqp/devdoc/sasl_tpm_requirements.md, line 28 at r1](https://reviewable.io:443/reviews/azure/azure-iot-sdk-node/212#-L3pVXiowOD6XaucuWh3:-L3pVXiowOD6XaucuWh4:b7qeyps) ([raw file](https://github.com/azure/azure-iot-sdk-node/blob/1104e870439d2345151b143ac00efad5122411e5/provisioning/transport/amqp/devdoc/sasl_tpm_requirements.md#L28)):*
> ```Markdown
> ## getChallengeResponseContent(challenge: any): Promise<Buffer>;
> 
> **SRS_NODE_PROVISIONING_AMQP_SASL_TPM_18_003: [** If `getChallengeResponse` is called with a 1 byte challenge, it shall resolve with the the intial response that was passed into the constructor. **]**
> ```

typo (initial)
Also, i don't remember that part of the protocol?

---

*[provisioning/transport/amqp/devdoc/sasl_tpm_requirements.md, line 34 at r1](https://reviewable.io:443/reviews/azure/azure-iot-sdk-node/212#-L3p_52q0P8rRJnVKhdv:-L3p_52q0P8rRJnVKhdw:b-z3qpfa) ([raw file](https://github.com/azure/azure-iot-sdk-node/blob/1104e870439d2345151b143ac00efad5122411e5/provisioning/transport/amqp/devdoc/sasl_tpm_requirements.md#L34)):*
> ```Markdown
> **SRS_NODE_PROVISIONING_AMQP_SASL_TPM_18_005: [** If `getChallengeResponse` is called with a first byte that has 11 in the most significant bits, it shall call the challenge callback with the full challenge buffer **]**
> 
> **SRS_NODE_PROVISIONING_AMQP_SASL_TPM_18_006: [** If `ChallengeResponseCallback` is called without passing an error, the final `getChallangeResponse` promise shall be resolved. **]**
> ```

typo (getChallEngeResponse)

---

*[provisioning/transport/amqp/devdoc/sasl_tpm_requirements.md, line 36 at r1](https://reviewable.io:443/reviews/azure/azure-iot-sdk-node/212#-L3p_7yHdVfUtUwdanBo:-L3p_7yHdVfUtUwdanBp:bmpnm1j) ([raw file](https://github.com/azure/azure-iot-sdk-node/blob/1104e870439d2345151b143ac00efad5122411e5/provisioning/transport/amqp/devdoc/sasl_tpm_requirements.md#L36)):*
> ```Markdown
> **SRS_NODE_PROVISIONING_AMQP_SASL_TPM_18_006: [** If `ChallengeResponseCallback` is called without passing an error, the final `getChallangeResponse` promise shall be resolved. **]**
> 
> **SRS_NODE_PROVISIONING_AMQP_SASL_TPM_18_007: [** If `ChallengeResponseCallback` is called with an error, the final `getChallangeResponse` promise shall be rejected. **]**
> ```

same typo here

---

*[provisioning/transport/amqp/src/amqp.ts, line 46 at r1](https://reviewable.io:443/reviews/azure/azure-iot-sdk-node/212#-L3pWHToGJJ_cPsYMJeW:-L3pWHToGJJ_cPsYMJeX:b-rvr0uv) ([raw file](https://github.com/azure/azure-iot-sdk-node/blob/1104e870439d2345151b143ac00efad5122411e5/provisioning/transport/amqp/src/amqp.ts#L46)):*
> ```XML
>   private _amqpStateMachine: machina.Fsm;
>   private _x509Auth: X509;
>   private _ek: Buffer;
> ```

nit: that's a little different from Tony's HTTP transport which uses a single object called _publicKeys I think (?) and contains the ek and srk. Also, same thing I said in Tony's PR, full variable names improve readability over abbreviations/acronyms :)

---

*[provisioning/transport/amqp/src/amqp.ts, line 158 at r1](https://reviewable.io:443/reviews/azure/azure-iot-sdk-node/212#-L3pXT68J-JVI_UmPLQQ:-L3pXT68J-JVI_UmPLQR:b-g9k23w) ([raw file](https://github.com/azure/azure-iot-sdk-node/blob/1104e870439d2345151b143ac00efad5122411e5/provisioning/transport/amqp/src/amqp.ts#L158)):*
> ```XML
>           respondToAuthenticationChallenge: (request, sasToken, callback) => {
>             let completionCompleteHandler = this._amqpStateMachine.on('tpmConnectionComplete', (err) => {
>               this._amqpStateMachine.off(completionCompleteHandler);
> ```

why not use standard EventEmitter APIs ? (asking out of curiosity)

---

*[provisioning/transport/amqp/src/amqp.ts, line 169 at r1](https://reviewable.io:443/reviews/azure/azure-iot-sdk-node/212#-L3pXtkQEBhSJyvoQU_2:-L3pXtkQEBhSJyvoQU_3:bjubkcd) ([raw file](https://github.com/azure/azure-iot-sdk-node/blob/1104e870439d2345151b143ac00efad5122411e5/provisioning/transport/amqp/src/amqp.ts#L169)):*
> ```XML
>           },
> 
>           tpmConnectionComplete: (err) => {
> ```

Isn't this pattern (with the internal 'tpmConnectionComplete' event) "emulating" an intermediate state that we are not describing? 

---

*[provisioning/transport/amqp/src/amqp.ts, line 500 at r1](https://reviewable.io:443/reviews/azure/azure-iot-sdk-node/212#-L3p_TjZG8Bc9YRTwJPg:-L3p_TjZG8Bc9YRTwJPh:bu9xqbm) ([raw file](https://github.com/azure/azure-iot-sdk-node/blob/1104e870439d2345151b143ac00efad5122411e5/provisioning/transport/amqp/src/amqp.ts#L500)):*
> ```XML
>   private _getAuthChallenge(request: RegistrationRequest, callback: (err: Error, tpmChallenge?: TpmChallenge) => void): void {
>     let hostname: string = request.idScope + '/registrations/' + request.registrationId;
>     let init: Buffer = new Builder()
> ```

I'm not sure this belongs here. IMHO the formatting of the response is the responsibility of the SaslTpm class, I would much rather see it handle this and just pass it an ek and srk instead of having to deal with "firstResponse" which, unless I fully understand the protocol, I'm not sure why it's separate from other challenges.

also that way the BufferBuilder and Promise dependencies are handled by the SaslTpm mechanism.

---

*[provisioning/transport/amqp/src/amqp.ts, line 530 at r1](https://reviewable.io:443/reviews/azure/azure-iot-sdk-node/212#-L3pbX-BqqvDPYQozbnb:-L3pbX-BqqvDPYQozbnc:b-ug5feg) ([raw file](https://github.com/azure/azure-iot-sdk-node/blob/1104e870439d2345151b143ac00efad5122411e5/provisioning/transport/amqp/src/amqp.ts#L530)):*
> ```XML
> 
>   private _respondToAuthChallenge(sasToken: string): void {
>     let responseBuffer: Buffer = new Builder()
> ```

same here. the way the buffer is structured imho belongs in the SaslTpm class.

---

*[provisioning/transport/amqp/src/sasl_tpm.ts, line 32 at r1](https://reviewable.io:443/reviews/azure/azure-iot-sdk-node/212#-L3paQdbkIj5zqu8pm6X:-L3paQdbkIj5zqu8pm6Y:b-bmd763) ([raw file](https://github.com/azure/azure-iot-sdk-node/blob/1104e870439d2345151b143ac00efad5122411e5/provisioning/transport/amqp/src/sasl_tpm.ts#L32)):*
> ```XML
>     this._init = init;
>     this._firstResponse = firstResponse;
>     this._getResponseFromChallenge = getResponseFromChallenge;
> ```

wouldn't that be getResponseForChallenge? also, in order to make it explicit for our case (instead of using generic AMQP words) wouldn't that be something like getSasToken?

---

*[provisioning/transport/amqp/src/sasl_tpm.ts, line 58 at r1](https://reviewable.io:443/reviews/azure/azure-iot-sdk-node/212#-L3pZ61LqE0HQq2yGJij:-L3pZ61LqE0HQq2yGJik:b6s4v3h) ([raw file](https://github.com/azure/azure-iot-sdk-node/blob/1104e870439d2345151b143ac00efad5122411e5/provisioning/transport/amqp/src/sasl_tpm.ts#L58)):*
> ```XML
>         /*Codes_SRS_NODE_PROVISIONING_AMQP_SASL_TPM_18_003: [ If `getChallengeResponse` is called with a 1 byte challenge, it shall resolve with the the intial response that was passed into the constructor. ] */
>         // initial challenge, reply with the srk
>         debug('initial response');
> ```

that's the one with the SRK right? I'd be explicit (at least in the naming) about it rather than just calling it "first" otherwise it looks weird why we'd separate the first response from the others.

---

*[provisioning/transport/amqp/test/_sasl_tpm_test.js, line 21 at r1](https://reviewable.io:443/reviews/azure/azure-iot-sdk-node/212#-L3pdxTVX40AWIqyzuIw:-L3pdxTVX40AWIqyzuIx:b-cpmheb) ([raw file](https://github.com/azure/azure-iot-sdk-node/blob/1104e870439d2345151b143ac00efad5122411e5/provisioning/transport/amqp/test/_sasl_tpm_test.js#L21)):*
> ```JavaScript
>   var fakeError = new Error('__FAKE_ERROR_TEXT__');
> 
>   var challangeFromBuffer = function(firstByte, buffer) {
> ```

typo (challEngeFromBuffer)

---

*[provisioning/transport/amqp/test/_sasl_tpm_test.js, line 55 at r1](https://reviewable.io:443/reviews/azure/azure-iot-sdk-node/212#-L3pecR5Uo0PQSHBfjnh:-L3pecR5Uo0PQSHBfjni:bg8jlnp) ([raw file](https://github.com/azure/azure-iot-sdk-node/blob/1104e870439d2345151b143ac00efad5122411e5/provisioning/transport/amqp/test/_sasl_tpm_test.js#L55)):*
> ```JavaScript
>     it ('resolves to the intial response on first call', function(callback) {
>       var sasl = new SaslTpm(fakeHostname, fakeInit, fakeFirstResponse, function() {});
>       sasl.getChallengeResponseContent(challangeFromBuffer(1, new Buffer('')))
> ```

probably needs to do a big search for that specific typo :) not going to point it out more.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/azure/azure-iot-sdk-node/212#-:-L3pTvWnU27A2I-dtYYx:b898fxl)*

File: provisioning/transport/amqp/src/amqp.ts [hunk 1]
```diff
@@ -8,11 +8,13 @@ import * as machina from 'machina';
 import * as dbg from 'debug';
 import * as uuid from 'uuid';
 import * as async from 'async';
+import * as Builder from 'buffer-builder';
 const debug = dbg('azure-iot-provisioning-device-amqp:Amqp');
 
 import { X509, errors } from 'azure-iot-common';
-import { ProvisioningTransportOptions, X509ProvisioningTransport, RegistrationRequest, RegistrationResult, ProvisioningDeviceConstants } from 'azure-iot-provisioning-device';
+import { ProvisioningTransportOptions, X509ProvisioningTransport, TpmProvisioningTransport, RegistrationRequest, RegistrationResult, ProvisioningDeviceConstants, TpmChallenge } from 'azure-iot-provisioning-device';
```
Comment (anthonyvercolano) [inline provisioning/transport/amqp/src/amqp.ts]: >, TpmChallenge [](start = 160, length = 14)

I killed this type.

File: provisioning/transport/amqp/src/sasl_tpm.ts [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    return new Promise((resolve: any) => {
+      resolve({
+        mechanism: this.name,
+        initialResponse: this._init,
+        hostname: this._hostname
+      });
+    });
+  }
+
+  // TODO: state machine here
```
Comment (anthonyvercolano) [inline provisioning/transport/amqp/src/sasl_tpm.ts]: >// TODO: state machine here [](start = 2, length = 27)

Coming in next review?

**Comment by BertKleewein** [comment]
Review status: all files reviewed at latest revision, 21 unresolved discussions.

---

*[common/transport/amqp/src/amqp.ts, line 570 at r1](https://reviewable.io:443/reviews/azure/azure-iot-sdk-node/212#-L3pMFNDKK1Pd12-0tGf:-L4CWZ3qkZNDxLuNkExX:b-896fix) ([raw file](https://github.com/azure/azure-iot-sdk-node/blob/1104e870439d2345151b143ac00efad5122411e5/common/transport/amqp/src/amqp.ts#L570)):*
<details><summary><i>Previously, pierreca (Pierre Cauchois) wrote…</i></summary><blockquote>

nit: setUriAndSocketTransport ? after all it's not only about the URI..
</blockquote></details>

Done.

---

*[common/transport/mqtt/package.json, line 11 at r1](https://reviewable.io:443/reviews/azure/azure-iot-sdk-node/212#-L3pLj-MAAZWGfe-hvKC:-L4808O5t-0Edvde0atB:b-3rywhz) ([raw file](https://github.com/azure/azure-iot-sdk-node/blob/1104e870439d2345151b143ac00efad5122411e5/common/transport/mqtt/package.json#L11)):*
<details><summary><i>Previously, pierreca (Pierre Cauchois) wrote…</i></summary><blockquote>

are we really  missing this or did this sneak up in the review inadvertantly?
</blockquote></details>

Really missing.  Only shows up if you're using an NPM that still nests node_modules.

---

*[provisioning/e2e/_provisioning_e2e.js, line 31 at r1](https://reviewable.io:443/reviews/azure/azure-iot-sdk-node/212#-L3pPXgJW6iRTw7pNu4C:-L4CWtJquSfbIsM2HtEZ:b-896fix) ([raw file](https://github.com/azure/azure-iot-sdk-node/blob/1104e870439d2345151b143ac00efad5122411e5/provisioning/e2e/_provisioning_e2e.js#L31)):*
<details><summary><i>Previously, pierreca (Pierre Cauchois) wrote…</i></summary><blockquote>

this is also going to avoid warning when removing tests and forgetting to remove require statements. it's not very important... why not rely on .skip() instead of commenting tests?
</blockquote></details>

Done.

---

*[provisioning/e2e/_provisioning_e2e.js, line 147 at r1](https://reviewable.io:443/reviews/azure/azure-iot-sdk-node/212#-L3pSeWQ7HDlg0YBttRn:-L4CXI4HjbL4kewimlNs:bkz5ckv) ([raw file](https://github.com/azure/azure-iot-sdk-node/blob/1104e870439d2345151b143ac00efad5122411e5/provisioning/e2e/_provisioning_e2e.js#L147)):*
<details><summary><i>Previously, pierreca (Pierre Cauchois) wrote…</i></summary><blockquote>

if you use this here, aren't you going to need to bind(this) later? why not use self since you seem to have it already?
</blockquote></details>

Done.  Good catch.

---

*[provisioning/transport/amqp/devdoc/sasl_tpm_requirements.md, line 18 at r1](https://reviewable.io:443/reviews/azure/azure-iot-sdk-node/212#-L3pUku-xiA77J3iWmqT:-L4CXYtu1ll1GLcXby3s:b-896fix) ([raw file](https://github.com/azure/azure-iot-sdk-node/blob/1104e870439d2345151b143ac00efad5122411e5/provisioning/transport/amqp/devdoc/sasl_tpm_requirements.md#L18)):*
<details><summary><i>Previously, pierreca (Pierre Cauchois) wrote…</i></summary><blockquote>

this has been renamed to getInitFrame()
</blockquote></details>

Done.

---

*[provisioning/transport/amqp/devdoc/sasl_tpm_requirements.md, line 26 at r1](https://reviewable.io:443/reviews/azure/azure-iot-sdk-node/212#-L3pUnZ8RwR8wD75cVn3:-L4CYvioVPpVBIHHoCbw:b-896fix) ([raw file](https://github.com/azure/azure-iot-sdk-node/blob/1104e870439d2345151b143ac00efad5122411e5/provisioning/transport/amqp/devdoc/sasl_tpm_requirements.md#L26)):*
<details><summary><i>Previously, pierreca (Pierre Cauchois) wrote…</i></summary><blockquote>

this has been renamed to getResponseFrame(). Also the result of the promise must be an object that contains a response field that contains the Buffer, in other words the return type is:
```ts
type SaslResponseFrame = { response: <Buffer> };
Promise<SaslResponseFrame>
```
</blockquote></details>

Done.

---

*[provisioning/transport/amqp/devdoc/sasl_tpm_requirements.md, line 28 at r1](https://reviewable.io:443/reviews/azure/azure-iot-sdk-node/212#-L3pVXiowOD6XaucuWh3:-L4CZWyQBuYa0iNmSMW0:b-jupoeb) ([raw file](https://github.com/azure/azure-iot-sdk-node/blob/1104e870439d2345151b143ac00efad5122411e5/provisioning/transport/amqp/devdoc/sasl_tpm_requirements.md#L28)):*
<details><summary><i>Previously, pierreca (Pierre Cauchois) wrote…</i></summary><blockquote>

typo (initial)
Also, i don't remember that part of the protocol?
</blockquote></details>

Done.  It's there.  I just got it from your sample.

---

*[provisioning/transport/amqp/devdoc/sasl_tpm_requirements.md, line 34 at r1](https://reviewable.io:443/reviews/azure/azure-iot-sdk-node/212#-L3p_52q0P8rRJnVKhdv:-L4CZgXYxnlYN9gy55Fg:b-896fix) ([raw file](https://github.com/azure/azure-iot-sdk-node/blob/1104e870439d2345151b143ac00efad5122411e5/provisioning/transport/amqp/devdoc/sasl_tpm_requirements.md#L34)):*
<details><summary><i>Previously, pierreca (Pierre Cauchois) wrote…</i></summary><blockquote>

typo (getChallEngeResponse)
</blockquote></details>

Done.

---

*[provisioning/transport/amqp/devdoc/sasl_tpm_requirements.md, line 36 at r1](https://reviewable.io:443/reviews/azure/azure-iot-sdk-node/212#-L3p_7yHdVfUtUwdanBo:-L4CZguaANNRAxeOeZov:b-896fix) ([raw file](https://github.com/azure/azure-iot-sdk-node/blob/1104e870439d2345151b143ac00efad5122411e5/provisioning/transport/amqp/devdoc/sasl_tpm_requirements.md#L36)):*
<details><summary><i>Previously, pierreca (Pierre Cauchois) wrote…</i></summary><blockquote>

same typo here
</blockquote></details>

Done.

---

*[provisioning/transport/amqp/src/amqp.ts, line 158 at r1](https://reviewable.io:443/reviews/azure/azure-iot-sdk-node/212#-L3pXT68J-JVI_UmPLQQ:-L4CatxbYJCTf9wJEXqO:b-lo0co8) ([raw file](https://github.com/azure/azure-iot-sdk-node/blob/1104e870439d2345151b143ac00efad5122411e5/provisioning/transport/amqp/src/amqp.ts#L158)):*
<details><summary><i>Previously, pierreca (Pierre Cauchois) wrote…</i></summary><blockquote>

why not use standard EventEmitter APIs ? (asking out of curiosity)
</blockquote></details>

Mostly to avoid creating a new EventEmitter object, but also to try something different and use something that appeared "more native" to the machina environment.  

---

*[provisioning/transport/amqp/src/amqp.ts, line 169 at r1](https://reviewable.io:443/reviews/azure/azure-iot-sdk-node/212#-L3pXtkQEBhSJyvoQU_2:-L4Cb5AE2Oz_j5lTeC7z:bbt2zy) ([raw file](https://github.com/azure/azure-iot-sdk-node/blob/1104e870439d2345151b143ac00efad5122411e5/provisioning/transport/amqp/src/amqp.ts#L169)):*
<details><summary><i>Previously, pierreca (Pierre Cauchois) wrote…</i></summary><blockquote>

Isn't this pattern (with the internal 'tpmConnectionComplete' event) "emulating" an intermediate state that we are not describing? 
</blockquote></details>

I don't think so.  I noodled over this for a while and tied my brain in too many knots.  The problem I have with this is that we start the connection in one handler (getAuthenticationChallenge), which is where we define the connection callback,  but when the connection is complete, we need to complete a different handler.  (respondToAuthenticationChallenge).  In other words, the AMQP connect() call spans two different states and two different handlers, and that spanning really F's everything up.  

I wonder if your idea to have respond() be a method on the challenge object could change this.  

---

*[provisioning/transport/amqp/src/sasl_tpm.ts, line 51 at r1](https://reviewable.io:443/reviews/azure/azure-iot-sdk-node/212#-L3odHHeW2y43k4Dd5nl-r1-51:-L4Ccprc3c-373-TmXlr:bqaffdl) ([raw file](https://github.com/azure/azure-iot-sdk-node/blob/1104e870439d2345151b143ac00efad5122411e5/provisioning/transport/amqp/src/sasl_tpm.ts#L51)):*
<details><summary><i>Previously, anthonyvercolano (Anthony V. Ercolano) wrote…</i></summary><blockquote>

>// TODO: state machine here [](start = 2, length = 27)

Coming in next review?
</blockquote></details>

comment removed.  I don't think we _need_ one.  It would be nice, but we don't need it.

---

*[provisioning/transport/amqp/test/_sasl_tpm_test.js, line 21 at r1](https://reviewable.io:443/reviews/azure/azure-iot-sdk-node/212#-L3pdxTVX40AWIqyzuIw:-L4CdCFz-SYxgS0ownOU:b-896fix) ([raw file](https://github.com/azure/azure-iot-sdk-node/blob/1104e870439d2345151b143ac00efad5122411e5/provisioning/transport/amqp/test/_sasl_tpm_test.js#L21)):*
<details><summary><i>Previously, pierreca (Pierre Cauchois) wrote…</i></summary><blockquote>

typo (challEngeFromBuffer)
</blockquote></details>

Done.

---

*[provisioning/transport/amqp/test/_sasl_tpm_test.js, line 55 at r1](https://reviewable.io:443/reviews/azure/azure-iot-sdk-node/212#-L3pecR5Uo0PQSHBfjnh:-L4CdDTMnX9qSzpZF_-K:b-896fix) ([raw file](https://github.com/azure/azure-iot-sdk-node/blob/1104e870439d2345151b143ac00efad5122411e5/provisioning/transport/amqp/test/_sasl_tpm_test.js#L55)):*
<details><summary><i>Previously, pierreca (Pierre Cauchois) wrote…</i></summary><blockquote>

probably needs to do a big search for that specific typo :) not going to point it out more.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/azure/azure-iot-sdk-node/212)*

**Pull request closed without merging**